# We Import the Modules That We Will Use

In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import re
import numpy as np

# We Import the CSV That We Will Use

In [2]:
sharks = pd.read_csv('data/sharkattack.csv', encoding = 'latin-1')

In [3]:
sharks

Case Number         Date    Year        Type    Country  \
0      2018.06.25  25-Jun-2018  2018.0     Boating        USA   
1      2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA   
2      2018.06.09  09-Jun-2018  2018.0     Invalid        USA   
3      2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA   
4      2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO   
...           ...          ...     ...         ...        ...   
25718         NaN          NaN     NaN         NaN        NaN   
25719         NaN          NaN     NaN         NaN        NaN   
25720         NaN          NaN     NaN         NaN        NaN   
25721         NaN          NaN     NaN         NaN        NaN   
25722          xx          NaN     NaN         NaN        NaN   

                  Area                        Location     Activity  \
0           California     Oceanside, San Diego County     Paddling   
1              Georgia  St. Simon Island, Glynn County     Standing   
2               Hawaii                    Habush, Oahu      Surfing   
3      New South Wales              Arrawarra Headland      Surfing   
4               Colima                        La Ticla  Free diving   
...                ...                             ...          ...   
25718              NaN                             NaN          NaN   
25719              NaN                             NaN          NaN   
25720              NaN                             NaN          NaN   
25721              NaN                             NaN          NaN   
25722              NaN                             NaN          NaN   

                  Name Sex   ...         Species   \
0          Julie Wolfe    F  ...      White shark   
1      Adyson McNeely     F  ...              NaN   
2          John Denges    M  ...              NaN   
3                 male    M  ...        2 m shark   
4       Gustavo Ramos     M  ...  Tiger shark, 3m   
...                ...  ...  ...              ...   
25718              NaN  NaN  ...              NaN   
25719              NaN  NaN  ...              NaN   
25720              NaN  NaN  ...              NaN   
25721              NaN  NaN  ...              NaN   
25722              NaN  NaN  ...              NaN   

               Investigator or Source                       pdf  \
0                    R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1      K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2      K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3                      B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4                           A .Kipper      2018.06.04-Ramos.pdf   
...                               ...                       ...   
25718                             NaN                       NaN   
25719                             NaN                       NaN   
25720                             NaN                       NaN   
25721                             NaN                       NaN   
25722                             NaN                       NaN   

                                            href formula  \
0      http://sharkattackfile.net/spreadsheets/pdf_di...   
1      http://sharkattackfile.net/spreadsheets/pdf_di...   
2      http://sharkattackfile.net/spreadsheets/pdf_di...   
3      http://sharkattackfile.net/spreadsheets/pdf_di...   
4      http://sharkattackfile.net/spreadsheets/pdf_di...   
...                                                  ...   
25718                                                NaN   
25719                                                NaN   
25720                                                NaN   
25721                                                NaN   
25722                                                NaN   

                                                    href Case Number.1  \
0      http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1      http://sharkattackfile.net/spreadsheets/pdf_di...    2

## Lets do some initial exploratory data analysis to understand our data set more


In [4]:
sharks.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

In [5]:
sharks.shape

(25723, 24)

We have over 25723 different entries of shark attacks


In [6]:
sharks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25723 entries, 0 to 25722
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             8702 non-null   object 
 1   Date                    6302 non-null   object 
 2   Year                    6300 non-null   float64
 3   Type                    6298 non-null   object 
 4   Country                 6252 non-null   object 
 5   Area                    5847 non-null   object 
 6   Location                5762 non-null   object 
 7   Activity                5758 non-null   object 
 8   Name                    6092 non-null   object 
 9   Sex                     5737 non-null   object 
 10  Age                     3471 non-null   object 
 11  Injury                  6274 non-null   object 
 12  Fatal (Y/N)             5763 non-null   object 
 13  Time                    2948 non-null   object 
 14  Species                 3464 non-null 

Most Columns are listed as strings, something that we will need to change in order to analyze our data 

In [7]:
sharks.isnull().sum()

Case Number               17021
Date                      19421
Year                      19423
Type                      19425
Country                   19471
Area                      19876
Location                  19961
Activity                  19965
Name                      19631
Sex                       19986
Age                       22252
Injury                    19449
Fatal (Y/N)               19960
Time                      22775
Species                   22259
Investigator or Source    19438
pdf                       19421
href formula              19422
href                      19421
Case Number.1             19421
Case Number.2             19421
original order            19414
Unnamed: 22               25722
Unnamed: 23               25721
dtype: int64

Most columns have over 17,000 NaN or missing values


# Defining Some Working Hypothesis

### H1 = Most shark attacks happened to male surfers younger than 40 years old
to answer this hypothesis we will need to work with the activity, age, and sex columns. 
### H2 = Most shark attacks where caused by great white sharks
to answer this hypothesis we will need to work with Species column
### H3 = Shark Attacks are correlated to an increase in climate change
to answer this hypothesis we will need to bring in another data set into the mix and find out if there is any correlation between climate change and shark attacks

# Creating Workable Data 

## We rename interesting columns that contain white spaces and weird spacing

In [8]:
sharks.rename(columns = {'Sex ': 'Sex', 'Species ': 'Species', 'Fatal (Y/N)':'Fatal_Y_N'}, inplace=True)
columns_h1 = ['Year','Age', 'Sex', 'Activity', 'Species', 'Country', 'Fatal_Y_N']


## creamos una lista de las columnas que nos interesan para la primera hypotesis

In [9]:
columns_h1 = ['Year','Age', 'Sex', 'Activity', 'Species', 'Country', 'Fatal_Y_N']
sharks_h1 = sharks[columns_h1]
sharks_h1

Year  Age  Sex     Activity          Species    Country Fatal_Y_N
0      2018.0   57    F     Paddling      White shark        USA         N
1      2018.0   11    F     Standing              NaN        USA         N
2      2018.0   48    M      Surfing              NaN        USA         N
3      2018.0  NaN    M      Surfing        2 m shark  AUSTRALIA         N
4      2018.0  NaN    M  Free diving  Tiger shark, 3m     MEXICO         N
...       ...  ...  ...          ...              ...        ...       ...
25718     NaN  NaN  NaN          NaN              NaN        NaN       NaN
25719     NaN  NaN  NaN          NaN              NaN        NaN       NaN
25720     NaN  NaN  NaN          NaN              NaN        NaN       NaN
25721     NaN  NaN  NaN          NaN              NaN        NaN       NaN
25722     NaN  NaN  NaN          NaN              NaN        NaN       NaN

[25723 rows x 7 columns]

## We delete all the columns that contain all missing values

In [10]:
sharks_h1.dropna(how = 'all', inplace = True)

<ipython-input-10-39ce4c7e3d07>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sharks_h1.dropna(how = 'all', inplace = True)


In [11]:
sharks_h1

Year  Age Sex      Activity          Species             Country  \
0     2018.0   57   F      Paddling      White shark                 USA   
1     2018.0   11   F      Standing              NaN                 USA   
2     2018.0   48   M       Surfing              NaN                 USA   
3     2018.0  NaN   M       Surfing        2 m shark           AUSTRALIA   
4     2018.0  NaN   M   Free diving  Tiger shark, 3m              MEXICO   
...      ...  ...  ..           ...              ...                 ...   
6297     0.0  NaN   M        Diving              NaN           AUSTRALIA   
6298     0.0  NaN   M  Pearl diving              NaN           AUSTRALIA   
6299     0.0  NaN   M      Swimming              NaN                 USA   
6300     0.0  NaN   M           NaN              NaN              PANAMA   
6301     0.0   15   M      Swimming              NaN  CEYLON (SRI LANKA)   

     Fatal_Y_N  
0            N  
1            N  
2            N  
3            N  
4            N  
...        ...  
6297         Y  
6298         Y  
6299         Y  
6300         Y  
6301         Y  

[6302 rows x 7 columns]

In [12]:
percent_left = (6302/25723) * 100
percent_left

24.499475177856393

We are left with around 24% of the orginal data when selecting columns and getting rid of the columns with all missing data

In [13]:
sharks_h1.Activity.value_counts()

Surfing                                                                   971
Swimming                                                                  869
Fishing                                                                   431
Spearfishing                                                              333
Bathing                                                                   162
                                                                         ... 
Hospital Ship Centaur torpedoed & sunk by the Japanese submarine I-177      1
Spearfishing, carrying his catch                                            1
Fell overboard from the Selim                                               1
Dynamite fishing                                                            1
Gloucester fishing schooner run down                                        1
Name: Activity, Length: 1532, dtype: int64

As we can see there is still a lot of NaN data left. Our first hypothesis focuses on the Activity, Age, Sex, and species columns so we will apply another Drop NAN to these columns in the hope of having a workable data set. 

In [14]:
sharks_h1_drop = sharks_h1.dropna(how='any', subset = ['Activity', 'Age', 'Sex', 'Species'])

In [15]:
sharks_h1_drop.Activity.value_counts()

Surfing                                        506
Swimming                                       305
Spearfishing                                   181
Fishing                                         80
Wading                                          71
                                              ... 
Spearfishing with Scuba gear                     1
Attempting to rescue a shark                     1
Checking fish traps, fell into the water         1
Swimming after boat became disabled              1
Swimming, after falling off towed kneeboard      1
Name: Activity, Length: 514, dtype: int64

In [16]:
sharks_h1_drop.shape

(2116, 7)

We are left with next to no missing values in the columns we are interested in. 
We also are left with 2116 rows of data

# Lets Start by Cleaning the Activity Column

In [17]:
sharks_h1_drop.Activity.unique()

array(['Paddling', 'Swimming', 'Fishing', 'Walking', 'Feeding sharks',
       'Scuba diving', 'Surfing', 'Paddle-skiing', 'Body boarding',
       'Stand-Up Paddleboarding', 'Wading', 'Scuba Diving',
       'Cleaning fish', 'Snorkeling', 'Spearfishing', 'Free diving',
       'Fishing / Wading', 'Night bathing', 'Kayaking / Fishing',
       'Body surfing', 'SUP', 'Kayaking', 'Surfing ', 'Touching a shark',
       'Attempting to lasso a shark', 'Washing hands', 'Surf skiing ',
       'Surf fishing', 'Floating', 'SUP Foil boarding', 'Lobstering',
       'Kite surfing', 'Floating in tube', 'Standing', 'Diving',
       'Teasing a shark', 'Feeding stingrays?', 'Paddle boarding',
       'Lobster fishing', 'Kayak Fishing', 'Surf-skiing',
       'Scallop diving on hookah', 'Treading water', 'Body boarding ',
       'Playing in the water', 'Body Boarding', 'Boogie boarding',
       'Playing', 'Attempting to rescue a shark',
       'Photographing the shark', 'Standing / Snorkeling',
       'Hand f

There are way to many not fitting values to our variable, we are going to try and extract as much information from the non standarized columns as we can. Lets start by creating a working copy of our subseted data frame to work on

In [18]:
sharks_copy = sharks_h1_drop.copy()

From SRC we import the funciones file which contains all our data cleaning functions

In [19]:
import SRC.funciones as fu

We create a list of regex terms and a list of what we want to change those regex terms too

In [20]:
list_regex = [r".*?\bswimming\b.*",r'.*?\bsurfing\b.*', r'.*?\bpaddling\b.*', r'.*?\bfishing\b.*', r'.*?\bspearfishing\b.*', r'.*?\bwading\b.*', r'.*?\bscuba\b\s+\bdiving\b.*', r'.*?\bsnorkeling\b.*', r'.*?\bwalking\b.*', r'.*?\bfree\b\s+\bdiving\b.*', r'.*?\bbody\b\s+\bboarding\b.*', r'.*?\bstanding\b.*', r'.*?\bbathing\b.*']
to_change = ['swimming', 'surfing', 'paddling', 'fishing', 'spearfishing', 'wading', 'scuba diving', 'snorkeling', 'walking', 'free diving', 'body boarding', 'standing', 'bathing']

We apply our replace_regex function to the activities column in order to extract the pertinent data from it and create a new column with said values

In [21]:
sharks3 = fu.replace_regex(sharks_copy, list_regex, to_change, 'Activity_Regex', 'Activity' )

As we can see the value counts for the top activities have increased after our data cleaning

In [22]:
sharks3.Activity_Regex.value_counts().head(10)

surfing          568
swimming         399
spearfishing     231
fishing          144
wading            77
scuba diving      72
snorkeling        55
standing          50
body boarding     42
free diving       38
Name: Activity_Regex, dtype: int64

# Lets Clean the Age Column

In order to clean the Age Column we will use a different function of the funciones.py to extract only the numbers from the age column and create a new column with them. We will start by creating a regex pattern that should encompass all numbers

In [23]:
number_regex = r'([0-9]+)'

In [24]:
sharks_age = fu.extract_regex(sharks3, 'Age_Regex', 'Age', number_regex)

In [25]:
sharks_age.head()

Year Age Sex        Activity          Species    Country Fatal_Y_N  \
0   2018.0  57   F        Paddling      White shark        USA         N   
6   2018.0  18   M        Swimming      Tiger shark     BRAZIL         Y   
7   2018.0  52   M         Fishing  Lemon shark, 3'        USA         N   
8   2018.0  15   M         Walking   Bull shark, 6'        USA         N   
10  2018.0  32   M  Feeding sharks  Grey reef shark  AUSTRALIA         N   

    Activity_Regex Age_Regex  
0         paddling        57  
6         swimming        18  
7          fishing        52  
8          walking        15  
10  feeding sharks        32

In [26]:
sharks_age.Age_Regex.value_counts().head(10)

19    90
18    83
20    82
17    80
21    80
22    77
15    75
16    74
25    73
24    69
Name: Age_Regex, dtype: int64

We turn the values in the age column to numeric values

In [27]:
pd.to_numeric(sharks_age['Age_Regex'])

0       57.0
6       18.0
7       52.0
8       15.0
10      32.0
        ... 
6239    36.0
6242     6.0
6243    16.0
6265    13.0
6276    16.0
Name: Age_Regex, Length: 2116, dtype: float64

# Lets Clean The Sex Column

As we previously did for activities, we will clean the sex column using the replace_regex function. We start by creating two lists


In [28]:
regex_sex = [r"m ", r".*?\blli\b.*"]
sex = ['m', 'NaN']

We apply our function

In [29]:
sharks_replace_sex = fu.replace_regex(sharks_age, regex_sex, sex, 'Sex_Regex', 'Sex')

In [30]:
sharks_replace_sex.nunique()

Year               115
Age                113
Sex                  3
Activity           514
Species           1069
Country            100
Fatal_Y_N            5
Activity_Regex     248
Age_Regex           73
Sex_Regex            3
dtype: int64

As we can see the Sex column is left with only 3 unique values, one of which is a NAN

# Lets Clean the Species Column


To clean the Species column we will be using our replace_regex function to extract the values that fit the species that we will be focusing on and pass them to a new column

In [31]:
species_regex = [r'.*?\bwhite\b\s+\bshark\b.*',r'.*?\bblue\b\s+\bshark\b.*', r'.*?\btiger\b\s.*',
           r'.*?\bbull\b\s.*',r'.*?\bshark\b\s+\binvolvement\b.*',r'.*?\bwobbegong\b\s+\bshark\b.*',
           r'.*?\bblacktip\b\s.*', r'.*?\bbronze\b\s+\bwhaler\b.*', r'.*?\bmako\b\s.*',r'.*?\bnurse\b\s.*',
           r'.*?\bhammerhead\b\s.*', r'.*?\braggedtooth\b\s.*']
species_change = ['White shark', 'Blue shark', 'Tiger shark', 
         'Bull shark', 'Not a shark', 'Wobbegong shark',
         'Blacktip shark','Bronze whaler shark', 'Mako shark', 
         'Nurse shark', 'Hammerhead shark', 'Raggedtooth shark']

In [32]:
Clean_Sharks = fu.replace_regex(sharks_replace_sex, species_regex, species_change, 'Species_Regex', 'Species')

In [33]:
Clean_Sharks.head(10)

Year Age Sex        Activity                Species     Country  \
0   2018.0  57   F        Paddling            White shark         USA   
6   2018.0  18   M        Swimming            Tiger shark      BRAZIL   
7   2018.0  52   M         Fishing        Lemon shark, 3'         USA   
8   2018.0  15   M         Walking         Bull shark, 6'         USA   
10  2018.0  32   M  Feeding sharks        Grey reef shark   AUSTRALIA   
13  2018.0  21   M         Fishing       Invalid incident     ENGLAND   
14  2018.0  34   F  Feeding sharks  Tawny nurse shark, 2m   AUSTRALIA   
18  2018.0  30   M    Scuba diving            Tiger shark  COSTA RICA   
21  2018.0  60   M         Surfing              3 m shark   AUSTRALIA   
23  2018.0  32   M         Fishing            Tiger shark    MALDIVES   

   Fatal_Y_N  Activity_Regex Age_Regex Sex_Regex     Species_Regex  
0          N        paddling        57         f       White shark  
6          Y        swimming        18         m       Tiger shark  
7          N         fishing        52         m   lemon shark, 3'  
8          N         walking        15         m        Bull shark  
10         N  feeding sharks        32         m   grey reef shark  
13         N         fishing        21         m  invalid incident  
14         N  feeding sharks        34         f       Nurse shark  
18         N    scuba diving        30         m       Tiger shark  
21         N         surfing        60         m         3 m shark  
23         N         fishing        32         m       Tiger shark

# Lets Clean the Fatal Column 

In [34]:
Clean_Sharks.Fatal_Y_N.unique()

array(['N', 'Y', nan, 'M', '2017', ' N'], dtype=object)

In [35]:
fatal_regex = [r'\s[a-zA-Z]+', r'm', r'2017']
fatal_change = ['n', np.nan, np.nan]

In [36]:
Clean_Sharks2 = fu.replace_regex(Clean_Sharks, fatal_regex, fatal_change, 'Fatal_Regex', 'Fatal_Y_N')

In [37]:
Clean_Sharks2.Fatal_Regex.unique()

array(['n', 'y', nan], dtype=object)

In [38]:
Clean_Sharks2

Year       Age Sex                                           Activity  \
0     2018.0        57   F                                           Paddling   
6     2018.0        18   M                                           Swimming   
7     2018.0        52   M                                            Fishing   
8     2018.0        15   M                                            Walking   
10    2018.0        32   M                                     Feeding sharks   
...      ...       ...  ..                                                ...   
6239     0.0        36   M                                         Skindiving   
6242     0.0         6   M  Swimming naked near a date palm where many dat...   
6243     0.0        16   M  Swimming in section of river used for washing ...   
6265     0.0  13 or 14   M                                           Swimming   
6276     0.0        16   M                                           Standing   

                 Species     Country Fatal_Y_N  Activity_Regex Age_Regex  \
0            White shark         USA         N        paddling        57   
6            Tiger shark      BRAZIL         Y        swimming        18   
7        Lemon shark, 3'         USA         N         fishing        52   
8         Bull shark, 6'         USA         N         walking        15   
10       Grey reef shark   AUSTRALIA         N  feeding sharks        32   
...                  ...         ...       ...             ...       ...   
6239  "A pack of sharks"  MOZAMBIQUE         N      skindiving        36   
6242          Bull shark        IRAQ         N        swimming         6   
6243          Bull shark        IRAQ         N        swimming        16   
6265          Bull shark        IRAQ         Y        swimming        13   
6276     12' tiger shark      BELIZE         Y        standing        16   

     Sex_Regex       Species_Regex Fatal_Regex  
0            f         White shark           n  
6            m         Tiger shark           y  
7            m     lemon shark, 3'           n  
8            m          Bull shark           n  
10           m     grey reef shark           n  
...        ...                 ...         ...  
6239         m  "a pack of sharks"           n  
6242         m          Bull shark           n  
6243         m          Bull shark           n  
6265         m          Bull shark           y  
6276         m         Tiger shark           y  

[2116 rows x 12 columns]

We go ahead and get rid of all NaN type values in our new clean data Frame

In [41]:
Clean_Sharks2.dropna(how='any', inplace= True)

In [42]:
Clean_Sharks2

Year       Age Sex                                           Activity  \
0     2018.0        57   F                                           Paddling   
6     2018.0        18   M                                           Swimming   
7     2018.0        52   M                                            Fishing   
8     2018.0        15   M                                            Walking   
10    2018.0        32   M                                     Feeding sharks   
...      ...       ...  ..                                                ...   
6239     0.0        36   M                                         Skindiving   
6242     0.0         6   M  Swimming naked near a date palm where many dat...   
6243     0.0        16   M  Swimming in section of river used for washing ...   
6265     0.0  13 or 14   M                                           Swimming   
6276     0.0        16   M                                           Standing   

                 Species     Country Fatal_Y_N  Activity_Regex Age_Regex  \
0            White shark         USA         N        paddling        57   
6            Tiger shark      BRAZIL         Y        swimming        18   
7        Lemon shark, 3'         USA         N         fishing        52   
8         Bull shark, 6'         USA         N         walking        15   
10       Grey reef shark   AUSTRALIA         N  feeding sharks        32   
...                  ...         ...       ...             ...       ...   
6239  "A pack of sharks"  MOZAMBIQUE         N      skindiving        36   
6242          Bull shark        IRAQ         N        swimming         6   
6243          Bull shark        IRAQ         N        swimming        16   
6265          Bull shark        IRAQ         Y        swimming        13   
6276     12' tiger shark      BELIZE         Y        standing        16   

     Sex_Regex       Species_Regex Fatal_Regex  
0            f         White shark           n  
6            m         Tiger shark           y  
7            m     lemon shark, 3'           n  
8            m          Bull shark           n  
10           m     grey reef shark           n  
...        ...                 ...         ...  
6239         m  "a pack of sharks"           n  
6242         m          Bull shark           n  
6243         m          Bull shark           n  
6265         m          Bull shark           y  
6276         m         Tiger shark           y  

[1927 rows x 12 columns]

We subset the data frame to only keep the new regex columns that we are interested in working with

In [43]:
CleanSharks = Clean_Sharks2[['Activity_Regex', 'Age_Regex', 'Sex_Regex', 'Species_Regex', 'Fatal_Regex']]

In [45]:
CleanSharks.reset_index()

index  Activity_Regex Age_Regex Sex_Regex       Species_Regex  \
0         0        paddling        57         f         White shark   
1         6        swimming        18         m         Tiger shark   
2         7         fishing        52         m     lemon shark, 3'   
3         8         walking        15         m          Bull shark   
4        10  feeding sharks        32         m     grey reef shark   
...     ...             ...       ...       ...                 ...   
1922   6239      skindiving        36         m  "a pack of sharks"   
1923   6242        swimming         6         m          Bull shark   
1924   6243        swimming        16         m          Bull shark   
1925   6265        swimming        13         m          Bull shark   
1926   6276        standing        16         m         Tiger shark   

     Fatal_Regex  
0              n  
1              y  
2              n  
3              n  
4              n  
...          ...  
1922           n  
1923           n  
1924           n  
1925           y  
1926           y  

[1927 rows x 6 columns]

In [48]:
CleanSharks.columns

Index(['Activity_Regex', 'Age_Regex', 'Sex_Regex', 'Species_Regex',
       'Fatal_Regex'],
      dtype='object')

# Our data is clean, time to export it 

In [49]:
CleanSharks.to_csv('data/CleanSharks.csv')